# Brief Intro about Named Entity Recognition (NER) using Spacy


In [1]:
import spacy

Docummentation about trained pipelines in spaCy https://spacy.io/models/en#en_core_web_sm

Download spaCy small (en_core_web_sm) and large (en_core_web_lg) trained pipeline.

Note: If this is your first time downloading it and you are in a Jupyter or Colab notebook, you may need to restart Python in order to load all the package's dependencies. You can do this by selecting the 'Restart kernel' or 'Restart runtime' option.

In [2]:
!python -m spacy download en_core_web_sm

Resolved 1 package in 451ms                                          
Installed 1 package in 3ms.0 (from https://github.com/explos
 + en-core-web-sm==3.8.0 (from https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.8.0/en_core_web_sm-3.8.0-py3-none-any.whl)
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


In [3]:
!python -m spacy download en_core_web_lg

Resolved 1 package in 339ms                                          
Installed 1 package in 5ms.0 (from https://github.com/explos
 + en-core-web-lg==3.8.0 (from https://github.com/explosion/spacy-models/releases/download/en_core_web_lg-3.8.0/en_core_web_lg-3.8.0-py3-none-any.whl)
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_lg')


# 1. Test the small pipeline

In [4]:
nlp = spacy.load("en_core_web_sm")
nlp

In [5]:
# test the pipeline
doc = nlp("Apple is looking at buying U.K. startup for $1 billion")

for token in doc:
    print("Token:",token.text, "Lemma:", token.lemma_, "POS:", token.pos_, token.tag_, token.dep_,
            token.shape_, token.is_alpha, token.is_stop)

Token: Apple Lemma: Apple POS: PROPN NNP nsubj Xxxxx True False
Token: is Lemma: be POS: AUX VBZ aux xx True True
Token: looking Lemma: look POS: VERB VBG ROOT xxxx True False
Token: at Lemma: at POS: ADP IN prep xx True True
Token: buying Lemma: buy POS: VERB VBG pcomp xxxx True False
Token: U.K. Lemma: U.K. POS: PROPN NNP nsubj X.X. False False
Token: startup Lemma: startup POS: VERB VBD ccomp xxxx True False
Token: for Lemma: for POS: ADP IN prep xxx True True
Token: $ Lemma: $ POS: SYM $ quantmod $ False False
Token: 1 Lemma: 1 POS: NUM CD compound d False False
Token: billion Lemma: billion POS: NUM CD pobj xxxx True False


Using trained pipeline from spaCy, you can extract linguistic features like token, lemma, Part of Speech (POS), POS tag, dependency, etc.

Now, let's use this spaCy pipeline for NER in the following texts:

In [6]:
# Text 1
doc1 = nlp("Victor Marie Hugo was a French poet, playwright, novelist, statesman and human rights activist.")
doc1

Victor Marie Hugo was a French poet, playwright, novelist, statesman and human rights activist.

In [7]:
from spacy import displacy # displacy helps us visualizing the entities

# Text 1
displacy.render(doc1, style="ent", jupyter=True)

In [8]:
# Text 2
doc2 = nlp("Pablo Ruiz Picasso was a Spanish painter, sculptor, printmaker, ceramicist, and theatre designer who spent most of his adult life in France.")
doc2

Pablo Ruiz Picasso was a Spanish painter, sculptor, printmaker, ceramicist, and theatre designer who spent most of his adult life in France.

In [9]:
# Text 2
displacy.render(doc2, style="ent", jupyter=True)

# 2. Text the large pipeline

In [10]:
nlp = spacy.load("en_core_web_lg")
nlp

In [11]:
doc = nlp("Pablo Ruiz Picasso was a Spanish painter, sculptor, printmaker, ceramicist, and theatre designer who spent most of his adult life in France.")
doc

Pablo Ruiz Picasso was a Spanish painter, sculptor, printmaker, ceramicist, and theatre designer who spent most of his adult life in France.

In [12]:
type(doc)

spacy.tokens.doc.Doc

In [13]:
# Check the recognized entities from the text
doc.ents

(Pablo Ruiz Picasso, Spanish, France)

In [14]:
doc.ents[0], type(doc.ents[0])

(Pablo Ruiz Picasso, spacy.tokens.span.Span)

In [15]:
from spacy import displacy
displacy.render(doc, style="ent", jupyter=True)

In [16]:
for ent in doc.ents:
    print(ent.text, ent.start_char, ent.end_char, ent.label_)

Pablo Ruiz Picasso 0 18 PERSON
Spanish 25 32 NORP
France 133 139 GPE


Entity description:

Value: 'Pablo Ruiz Picasso' --> Label: 'Person' --> Value position in text: (start:0, end:18)
Value: 'Spanish' --> Label: 'NORP' --> Value position in text: (start:25, end:32)
Value: 'France' --> Label: 'GPE' --> Value position in text: (start:133, end:139)

# Custom NER for Medical texts
IMPORTANT NOTE: Each dataset has its own structure, so before you start working, you must ensure that the dataset has all the necessary features, variables and information, and that it has the required data structure for spaCy.

We will use a dataset from https://www.kaggle.com/datasets/finalepoch/medical-ner

You can manually download the dataset from the link above or automatically download it using your Kaggle API key.

In [3]:
pip install opendatasets

Note: you may need to restart the kernel to use updated packages.


In [5]:
import opendatasets as od

ModuleNotFoundError: No module named 'cgi'